### PR TITLE
Add convenience methods to suggest the next subject or session number.

### PR DESCRIPTION
**Pull request by JoeZiminski**
This PR adds the convenience functions to next the next subject or session in the project. This required some re-working of an old method that retrieved the subject or session IDs, which is now called `utils/get_values_from_bids_formatted_name()`.

1) searches the local and remote repositories for all folders starting with `sub-*` (when finding next subject) or `ses-*` from within a subject directory (when finding next session). The core function is called `folders/get_next_sub_or_ses_number`.

2) parses the numerical value for subject / session. If these are not increasing by 1 each time, a warning is shown.

3) From the array of subject or session numbers, add 1 to suggest the next subject / session. API and CLI functions are provided (in `datashuttle.py` and `command_line_interface.py` respectively.

### CODE REVIEW

File: datashuttle/command_line_interface.py [hunk 1]
```diff
@@ -819,6 +837,32 @@ def construct_parser():
     )
     show_local_tree_parser.set_defaults(func=show_local_tree)
 
+    # Show Local tree
+    # -------------------------------------------------------------------------
+
+    show_next_sub_number_parser = subparsers.add_parser(
+        "show-next-sub-number",
+        aliases=["show_next_sub_number"],
+        description=process_docstring(DataShuttle.show_local_tree.__doc__),
```
Comment (adamltyson) [inline datashuttle/command_line_interface.py:846]: Should this be:
```suggestion
        description=process_docstring(DataShuttle.show_next_sub_number.__doc__),
```

File: datashuttle/command_line_interface.py [hunk 2]
```diff
@@ -819,6 +837,32 @@ def construct_parser():
...
+    )
+    show_next_sub_number_parser.set_defaults(func=show_next_sub_number)
+
+    # Show Local tree
+    # -------------------------------------------------------------------------
+
+    show_next_ses_number_parser = subparsers.add_parser(
+        "show-next-ses-number",
+        aliases=["show_next_ses_number"],
+        description=process_docstring(DataShuttle.show_local_tree.__doc__),
```
Comment (adamltyson) [inline datashuttle/command_line_interface.py:856]:
```suggestion
        description=process_docstring(DataShuttle.show_next_ses_number.__doc__),
```

File: datashuttle/command_line_interface.py [hunk 3]
```diff
@@ -819,6 +837,32 @@ def construct_parser():
...
     show_local_tree_parser.set_defaults(func=show_local_tree)
 
+    # Show Local tree
+    # -------------------------------------------------------------------------
+
+    show_next_sub_number_parser = subparsers.add_parser(
+        "show-next-sub-number",
+        aliases=["show_next_sub_number"],
+        description=process_docstring(DataShuttle.show_local_tree.__doc__),
+    )
```
Comment (adamltyson) [inline datashuttle/command_line_interface.py:847]:
```suggestion
        help="",
    )
```

File: datashuttle/command_line_interface.py [hunk 4]
```diff
@@ -819,6 +837,32 @@ def construct_parser():
...
+    show_next_sub_number_parser.set_defaults(func=show_next_sub_number)
+
+    # Show Local tree
+    # -------------------------------------------------------------------------
+
+    show_next_ses_number_parser = subparsers.add_parser(
+        "show-next-ses-number",
+        aliases=["show_next_ses_number"],
+        description=process_docstring(DataShuttle.show_local_tree.__doc__),
+    )
```
Comment (adamltyson) [inline datashuttle/command_line_interface.py:857]:
```suggestion
        help="",
    )
```

File: datashuttle/utils/folders.py [hunk 5]
```diff
@@ -463,3 +470,91 @@ def search_filesystem_path_for_folders(
...
+    Should this search filenames also?
+
+    """
+    if sub:
+        bids_key = "ses"
+    else:
+        bids_key = "sub"
+
+    # Search local and remote for folders that begin with "sub-*"
+    (local_foldernames, __,) = search_sub_or_ses_level(
```
Comment (adamltyson) [inline datashuttle/utils/folders.py:522]: In here and line 529, is there a reason to use a double underscore? I always used a single to signify an unused variable. Does `__` mean something different?

Also, are the parentheses needed?

File: datashuttle/datashuttle.py [hunk 6]
```diff
@@ -821,6 +839,47 @@ def show_local_tree(self):
         """
         ds_logger.print_tree(self.cfg["local_path"])
 
+    def show_next_sub_number(self) -> None:
+        """
+        Show a suggested value for the next available subject number
+        The local and remote repository will be searched, and the
+        maximum existing subject number + 1 will be suggested.
+
+        Note that this cannot search all local machines except for the one
```
Comment (adamltyson) [inline datashuttle/datashuttle.py:848]: I would clarify this docstring.

File: datashuttle/utils/folders.py [hunk 7]
```diff
@@ -463,3 +470,91 @@ def search_filesystem_path_for_folders(
...
+    cfg : datashuttle configs class
+    sub : subject name to search within if searching for sessions, otherwise None
+          to search for subjects
+    search_str : the string to search for within the top-level or subject-levle
+                 folder ("sub-*") or ("ses-*") are suggested, respectively.
+
+    Returns
+    -------
+    suggested_new_num : the new suggested sub / ses number.
+    latest_existing_num : the latest sub / ses number that currents
```
Comment (adamltyson) [inline datashuttle/utils/folders.py:509]:
```suggestion
    latest_existing_num : the latest sub / ses number that currently
```

File: datashuttle/utils/utils.py [hunk 8]
```diff
@@ -121,19 +123,68 @@ def log_and_raise_error_not_exists_or_not_yaml(path_to_config: Path) -> None:
...
-        for name in all_names
-        if re.split("-|_", name)[0] in ["sub", "ses"]
-    ]
+    all_values = []
+    for name in all_names:
+
+        if key not in name:
+            raise_error(f"They key {key} is not found in {name}")
+
+        value = re.findall(f"{key}-(.*?)(?=_|$)", name)
```
Comment (adamltyson) [inline datashuttle/utils/utils.py:165]: Could this regular expression be put into it's own function to enable it to be tested individually. Just thinking we may not notice if it changed slightly. Hopefully other tests would pick it up, but better safe than sorry.

**Review by adamltyson — APPROVED** [review]
Looks good. I left a few tiny suggestions, inc one that I _think_ would fix the tests (not 100% sure, I learnt some stuff reviewing this!).

File: datashuttle/utils/folders.py (hunk 5 above)
Comment (JoeZiminski) [reply datashuttle/utils/folders.py:522]: Funnily enough I used doubles based on some SO answers I read early when learning to code such as [this ](https://softwareengineering.stackexchange.com/a/285304), that indicate `_` can conflict with `gettext`. Having no idea what `gettext` was, I used `__` by default to be safe. However it appears this is unecessary and `_` is standard (unless your specifically using `gettext` in the codebase. Will change across the codebase in a new PR.

Not sure about the brackets I think this is some `black` magic, I'll check.

**Comment by JoeZiminski** [comment]
Cheers @adamltyson! Those fixes did make the test pass cheers, there's two requested changes it would be great to get your opinion on (in comments), then I think good to go! thanks

**Pull request merged**